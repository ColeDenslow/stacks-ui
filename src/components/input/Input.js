import React, { Component } from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';

import LoadingSpinner from '../loadingspinner/LoadingSpinner';

import './Input.css';

class Input extends Component {
  constructor(props) {
    super(props);

    this.state = {
      hasFocus: false,
    };

    this.fieldWrapperHandleClick = this.fieldWrapperHandleClick.bind(this);
    this.fieldWrapperHandleDoubleClick = this.fieldWrapperHandleDoubleClick.bind(this);

    this.onFocus = this.onFocus.bind(this);
    this.onBlur = this.onBlur.bind(this);
  }

  onFocus(evt) {
    this.setState({ hasFocus: true });

    if (this.props.onFocus) {
      this.props.onFocus(evt);
    }

    if (this.props.selectOnFocus === true) {
      this.inputRef.select();
    }
  }

  onBlur(evt) {
    this.setState({ hasFocus: false });

    if (this.props.onBlur) {
      this.props.onBlur(evt);
    }
  }

  fieldWrapperHandleClick() {
    this.inputRef.focus();
  }

  fieldWrapperHandleDoubleClick() {
    this.inputRef.select();
  }

  render() {
    const {
      className, disabled, loading, messageType, prefix, readOnly, selectOnFocus, size, suffix, type, ...otherProps
    } = this.props;

    const classNames = cn([
      'stx-input',
      { 'stx-input--is-disabled': disabled },
      { 'stx-input--is-loading': loading },
      { 'stx-input--is-read-only': readOnly },
      { 'stx-input--has-focus': this.state.hasFocus },
      { 'stx-input--alert': messageType === 'alert' || messageType === 'error' },
      { 'stx-input--warning': messageType === 'warning' },
      { 'stx-input--success': messageType === 'success' },
      { 'stx-input--small': size === 'small' },
      this.props.className,
    ]);

    return (
      <div
        className={classNames}
        onClick={this.fieldWrapperHandleClick}
        onDoubleClick={this.fieldWrapperHandleDoubleClick}
      >
        { prefix && <div className='stx-input__prefix'>{prefix}</div> }

        <input
          {...otherProps}
          ref={(input) => {
            this.inputRef = input;
          }}
          type={type}
          className='stx-input__input-element'
          disabled={disabled}
          readOnly={readOnly}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          aria-disabled={disabled}
          aria-readonly={readOnly}
        />

        { suffix && <div className='stx-input__suffix'>{suffix}</div> }

        { loading && <LoadingSpinner className='stx-input__spinner' /> }
      </div>
    );
  }
}

Input.defaultProps = {
  selectOnFocus: false,
  type: 'text',
};

Input.propTypes = {
  /** Optional className for the Input component */
  className: PropTypes.string,

  /** A node for placing read only content before the input element */
  prefix: PropTypes.node,
  /** A node for placing read only content after the input element */
  suffix: PropTypes.node,

  /** The input field type */
  type: PropTypes.oneOf([
    'text',
    'date',
    'datetime-local',
    'time',
    'search',
    'number',
    'tel',
    'email',
    'url',
    'password',
    'file',
  ]),

  /** Size modifier */
  size: PropTypes.oneOf(['small', '']),

  /** The type of message to display */
  messageType: PropTypes.oneOf(['alert', 'warning', 'success', '']),

  /** Whether or not the field is disabled */
  disabled: PropTypes.bool,

  /** Whether or not the field is read only */
  readOnly: PropTypes.bool,

  /** Whether or not the input field is in a loading state */
  loading: PropTypes.bool,

  /** Whether or not the value of the field should be selected when the field receives focus */
  selectOnFocus: PropTypes.bool,

  /** Event fires when the input receives focus */
  onFocus: PropTypes.func,

  /** Event fires when the input loses focus */
  onBlur: PropTypes.func,
};

export default Input;
