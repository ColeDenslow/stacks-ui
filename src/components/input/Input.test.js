/* global describe, test, expect */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import React from 'react';
import Input from './Input';

describe('Input', () => {
  test('Input renders with basic props', () => {
    const onFocus = () => 'onFocus';
    const onBlur = () => 'onBlur';
    const onChange = () => 'onChange';
    const onKeyUp = () => 'onKeyUp';
    const onKeyDown = () => 'onKeyDown';

    const output = shallow(<Input
      id='test-id'
      className='html-class'
      name='name'
      value='test-string'
      placeholder='placeholder'
      data-additional-prop='additional-prop-test'
      type='text'
      onFocus={onFocus}
      onBlur={onBlur}
      onChange={onChange}
      onKeyUp={onKeyUp}
      onKeyDown={onKeyDown}
      selectOnFocus
    />);

    expect(toJson(output)).toMatchSnapshot();
    expect(output.find('input').prop('onChange')).toEqual(onChange);
    expect(output.find('input').prop('onKeyUp')).toEqual(onKeyUp);
    expect(output.find('input').prop('onKeyDown')).toEqual(onKeyDown);
  });

  test('Input renders without any props', () => {
    const output = shallow(<Input />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with a prefix', () => {
    const output = shallow(<Input prefix='-' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with a suffix', () => {
    const output = shallow(<Input suffix='px' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with type text', () => {
    const output = shallow(<Input type='text' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with type date', () => {
    const output = shallow(<Input type='date' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with type datetime-local', () => {
    const output = shallow(<Input type='datetime-local' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with type time', () => {
    const output = shallow(<Input type='time' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with type search', () => {
    const output = shallow(<Input type='search' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with type number', () => {
    const output = shallow(<Input type='number' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with type tel', () => {
    const output = shallow(<Input type='tel' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with type email', () => {
    const output = shallow(<Input type='email' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with type url', () => {
    const output = shallow(<Input type='url' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with type password', () => {
    const output = shallow(<Input type='password' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with type file', () => {
    const output = shallow(<Input type='file' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with size small', () => {
    const output = shallow(<Input size='small' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with a message type of alert', () => {
    const output = shallow(<Input messageType='alert' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with a message type of warning', () => {
    const output = shallow(<Input messageType='warning' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with a message type of success', () => {
    const output = shallow(<Input messageType='success' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with a min and max value', () => {
    const output = shallow(<Input min={1} max={5} />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with a min and max length', () => {
    const output = shallow(<Input minLength={1} maxLength={5} />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with a step value', () => {
    const output = shallow(<Input step={5} />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with a tabIndex value', () => {
    const output = shallow(<Input tabIndex={-5} />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input renders with a autoComplete value', () => {
    const output = shallow(<Input autoComplete='off' />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input render disabled when disabled is true', () => {
    const output = shallow(<Input disabled />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input render readOnly when readOnly is true', () => {
    const output = shallow(<Input readOnly />);
    expect(toJson(output)).toMatchSnapshot();
  });

  test('Input render with a loading spinner when loading is true', () => {
    const output = shallow(<Input loading />);
    expect(toJson(output)).toMatchSnapshot();
  });
});
