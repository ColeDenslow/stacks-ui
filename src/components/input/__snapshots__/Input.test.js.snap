// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Input render disabled when disabled is true 1`] = `
<div
  className="stx-input stx-input--is-disabled"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    aria-disabled={true}
    className="stx-input__input-element"
    disabled={true}
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`Input Input render readOnly when readOnly is true 1`] = `
<div
  className="stx-input stx-input--is-read-only"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    aria-readonly={true}
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    readOnly={true}
    type="text"
  />
</div>
`;

exports[`Input Input render with a loading spinner when loading is true 1`] = `
<div
  className="stx-input stx-input--is-loading"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
  <LoadingSpinner
    className="stx-input__spinner"
  />
</div>
`;

exports[`Input Input renders with a autoComplete value 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    autoComplete="off"
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`Input Input renders with a message type of alert 1`] = `
<div
  className="stx-input stx-input--alert"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`Input Input renders with a message type of success 1`] = `
<div
  className="stx-input stx-input--success"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`Input Input renders with a message type of warning 1`] = `
<div
  className="stx-input stx-input--warning"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`Input Input renders with a min and max length 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    maxLength={5}
    minLength={1}
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`Input Input renders with a min and max value 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    max={5}
    min={1}
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`Input Input renders with a prefix 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <div
    className="stx-input__prefix"
  >
    -
  </div>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`Input Input renders with a step value 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    step={5}
    type="text"
  />
</div>
`;

exports[`Input Input renders with a suffix 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
  <div
    className="stx-input__suffix"
  >
    px
  </div>
</div>
`;

exports[`Input Input renders with a tabIndex value 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    tabIndex={-5}
    type="text"
  />
</div>
`;

exports[`Input Input renders with basic props 1`] = `
<div
  className="stx-input html-class"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    data-additional-prop="additional-prop-test"
    id="test-id"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    placeholder="placeholder"
    type="text"
    value="test-string"
  />
</div>
`;

exports[`Input Input renders with size small 1`] = `
<div
  className="stx-input stx-input--small"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`Input Input renders with type date 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="date"
  />
</div>
`;

exports[`Input Input renders with type datetime-local 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="datetime-local"
  />
</div>
`;

exports[`Input Input renders with type email 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="email"
  />
</div>
`;

exports[`Input Input renders with type file 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="file"
  />
</div>
`;

exports[`Input Input renders with type number 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="number"
  />
</div>
`;

exports[`Input Input renders with type password 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="password"
  />
</div>
`;

exports[`Input Input renders with type search 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="search"
  />
</div>
`;

exports[`Input Input renders with type tel 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="tel"
  />
</div>
`;

exports[`Input Input renders with type text 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`Input Input renders with type time 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="time"
  />
</div>
`;

exports[`Input Input renders with type url 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="url"
  />
</div>
`;

exports[`Input Input renders without any props 1`] = `
<div
  className="stx-input"
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <input
    className="stx-input__input-element"
    onBlur={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;
